runner/app/live/streamer/process.py

```diff
@@ -122,93 +124,118 @@ def process_loop(self):
-                            logging.info(f"PipelineProcess: Successfully applied params to pipeline: {params}")
-                    except Exception as e:
-                        error_msg = f"Error updating params: {str(e)}"
-                        logging.error(error_msg, exc_info=True)
-                        report_error(error_msg)
+                await asyncio.sleep(0.1)
 
-                try:
-                    input_frame = self.input_queue.get(timeout=0.1)
-                except queue.Empty:
-                    continue
+        tasks = [input_task, output_task, param_task, asyncio.create_task(wait_for_stop())]
 
-                try:
-                    if isinstance(input_frame, VideoFrame):
-                        input_frame.log_timestamps["pre_process_frame"] = time.time()
-                        output_image = pipeline.process_frame(input_frame.image)
-                        input_frame.log_timestamps["post_process_frame"] = time.time()
-                        output_frame = VideoOutput(input_frame.replace_image(output_image), self.request_id)
-                        self.output_queue.put(output_frame)
-                    elif isinstance(input_frame, AudioFrame):
-                        self.output_queue.put(AudioOutput([input_frame], self.request_id))
-                        # TODO wire in a proper pipeline here
-                    else:
-                        report_error(f"Unsupported input frame type {type(input_frame)}")
-                except Exception as e:
-                    report_error(f"Error processing frame: {e}")
+        try:
+            await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
         except Exception as e:
-            report_error(f"Error in process run method: {e}")
+            self._report_error(f"Error in pipeline loops: {e}")
         finally:
-            self._cleanup_pipeline(pipeline)
+            for task in tasks:
+                task.cancel()
+            await asyncio.gather(*tasks, return_exceptions=True)
+            await self._cleanup_pipeline(pipeline)
+
+    async def _input_loop(self, pipeline):
+        while not self.is_done():
+            try:
+                input_frame = await asyncio.to_thread(self.input_queue.get)
+                if isinstance(input_frame, VideoFrame):
+                    input_frame.log_timestamps["pre_process_frame"] = time.time()
+                    await pipeline.put_video_frame(input_frame)
+                elif isinstance(input_frame, AudioFrame):
+                    await asyncio.to_thread(self.output_queue.put, AudioOutput([input_frame], self.request_id))
+            except queue.Empty:
+                continue
+            except Exception as e:
+                self._report_error(f"Error processing input frame: {e}")
+
+    async def _output_loop(self, pipeline):
+        while not self.is_done():
+            try:
+                output_frame = await pipeline.get_processed_video_frame(self.request_id)
```

Comment:
Would there be a way to tag the images with a request_id on the input side instead? Otherwise I worry that we'll end up having the same issue we had before with the queues from the ai-runner code: frames from previous streams leaking to the next stream because of these queues, causing encoder/decoder crashes.

So if there could be a way we could "tag" the frame on the [input side on `put_video_frame`](https://github.com/livepeer/ai-runner/pull/479/files#diff-884ec47d03a7fdce1ff7a0fc4f9858f45965ff1cbb3ee034c1f3505f756a7f2bR198) we could actually track where the frame is coming from. Do you think that could work easily? Maybe on that `side_data` field. If it's not a simple change don't worry, this is more like a nice-to-have.

Reply:
I was concerned about that too, also considered using `side_data` to pass around the `request_id`, I read that it needed to remain consistent with the av frame type, wasn't fully sure if this comment is referring to VideoFrame or something else. Maybe we can create an issue for the backlog and @varshith15 can shed some light on this constraint?

Reply:
What if instead of putting `VideoFrame`s on the `video_incoming_frames` we put `VideoOutput` objects instead? Then we could have the request_id in them passed through the queue here